valohai_cli/utils/ssh.py

```diff
@@ -0,0 +1,147 @@
+    for pth in folder_path.glob("*.pub"):
+        if not pth.is_file():
+            continue
+        try:
+            with pth.open("rb") as f:
+                first_line = f.readline().strip()
+                if first_line.startswith((b"ssh-rsa", b"ecdsa-sha2-", b"ssh-ed25519")):
+                    private_key_paths = [pth.with_suffix(""), pth.with_suffix(".pem")]
+                    for private_key_path in private_key_paths:
+                        if private_key_path.is_file():
+                            yield Keypair(public_path=pth, private_path=private_key_path)
+                            break
+        except OSError:  # e.g. unreadable key
+            pass
+
+
+def select_private_key_from_possible_directories() -> Optional[str]:
+    default_ssh_dir = os.path.expanduser("~/.ssh")
+    cli_config_dir = get_settings_file_name("")
+    all_keypairs: list[Keypair] = []
+    all_dirs = [cli_config_dir, default_ssh_dir]
+    for dir in all_dirs:
+        if not os.path.isdir(dir):
+            continue
+        all_keypairs.extend(get_ssh_key_pairs_from_dir(dir))
+
+    if not all_keypairs:
+        raise click.UsageError(
+            f"No SSH key pairs found in {', '.join(all_dirs)}. Please try to generate new SSH key pair",
+        )
+
+    keypair_options = [
+        {
+            "name": keypair.private_path,
+            "description": f"Public: {keypair.public_path}",
+        }
+        for keypair in all_keypairs
+    ]
+
+    selected_key = prompt_from_list(prompt="Select the SSH key pair you want to use", options=keypair_options)
+    return str(selected_key["name"])
+
+
+def fetch_status_events(execution: dict, first_n: Optional[int] = None) -> Generator:
+    params = {}
+    if first_n is not None:
+        params["first_n"] = first_n
+    events_response = request("get", f"{execution['url']}status_events/", params=params).json()
+    yield from events_response.get("status_events", [])
+
+
+def get_execution_from_counter_with_retry(counter: int) -> dict:
+    project = get_project(require=True)
+    assert project
+
+    execution = call_with_retry(
+        func=lambda: project.get_execution_from_counter(
+            raise_on_not_running=True,
+            counter=counter,
+            params={
+                "exclude": "metadata, events",
+            },
+        ),
+        retry_on_exception_classes=(ExecutionNotRunning,),
+        retries=5,
+        delay_range=(1, 7),
+    )
+    return execution
+
+
+def get_ssh_details_from_execution(execution: dict) -> Tuple[str, int]:
+    events_response = fetch_status_events(execution, first_n=100)
+    ssh_details = find_ssh_details(list(events_response))
+    if not ssh_details:
+        raise click.UsageError("No SSH details found")
+    return ssh_details
+
+
+def get_ssh_details_with_retry(
+    counter: int,
+) -> Tuple[str, int]:
+    """
+    Fetch SSH details for a given execution, retrying up to `retries` times with a random delay between retries.
```

Comment:
The docstring is referring to now-deleted arguments...
```suggestion
    Fetch SSH details for a given execution.
```